controllers/apim/authpolicy_controller.go

```diff
@@ -0,0 +1,211 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// AuthPolicyReconciler reconciles a AuthPolicy object
+type AuthPolicyReconciler struct {
+	*reconcilers.BaseReconciler
+	Scheme *runtime.Scheme
+}
+
+const authPolicyFinalizer = "authpolicy.kuadrant.io/finalizer"
+
+//+kubebuilder:rbac:groups=apim.kuadrant.io,resources=authpolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apim.kuadrant.io,resources=authpolicies/finalizers,verbs=update
+//+kubebuilder:rbac:groups=security.istio.io,resources=authorizationpolicies,verbs=get;list;watch;create;update;patch;delete
+
+func (r *AuthPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger().WithValues("AuthPolicy", req.NamespacedName)
+	logger.Info("Reconciling AuthPolicy")
+	ctx := logr.NewContext(eventCtx, logger)
+
+	var ap apimv1alpha1.AuthPolicy
+	if err := r.Client().Get(ctx, req.NamespacedName, &ap); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("no AuthPolicy found")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get AuthPolicy")
+		return ctrl.Result{}, err
+	}
+
+	if ap.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(&ap, authPolicyFinalizer) {
+		logger.V(1).Info("Handling removal of authpolicy object")
+		if err := r.removeIstioAuthPolicy(ctx, &ap); err != nil {
+			logger.Error(err, "failed to remove Istio's AuthorizationPolicy")
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(&ap, authPolicyFinalizer)
+		if err := r.UpdateResource(ctx, &ap); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
+	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
+	if ap.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(&ap, authPolicyFinalizer) {
+		controllerutil.AddFinalizer(&ap, authPolicyFinalizer)
+		if err := r.UpdateResource(ctx, &ap); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	if err := r.reconcileAuthPolicy(ctx, &ap); err != nil {
+		logger.Error(err, "failed to reconcile AuthPolicy")
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("completed reconciling AuthPolicy")
+	return ctrl.Result{}, nil
+}
+
+// IstioAuthPolicy generates Istio's AuthorizationPolicy using Kuadrant's AuthPolicy
+func (r *AuthPolicyReconciler) reconcileAuthPolicy(ctx context.Context, ap *apimv1alpha1.AuthPolicy) error {
+	logger := logr.FromContext(ctx)
+
+	httpRoute, err := r.fetchHTTPRoute(ctx, ap)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
```

Comment:
if HTTPRoute is not found, shouldn't the controller at least retry?

Reply:
There is a possibility of a loop right then? when HTTPRoute won't exist even in the future.